plugins/ccgx/fu-ccgx-dock-bb.c

```diff
@@ -0,0 +1,734 @@
+
+		fu_device_set_status (device,FWUPD_STATUS_DEVICE_READ);
+
+		if (!fu_ccgx_hpi_cmd_read_flash (device, hpi_handle, row_num, row_data, row_size, error)) {
+			g_warning("fw meta data read error ");
+			return FALSE;
+		}
+
+		/* get meta data */
+		metadata = (CCGxMetaData*)&row_data[self->pd_device_data.fw_meta_offset];
+
+		/* clear meta data Valid */
+		metadata->metadata_valid = 0x00;
+
+		/* write meta data again */
+		fu_device_set_status (device,FWUPD_STATUS_DEVICE_ERASE);
+
+		if (!fu_ccgx_hpi_cmd_write_flash (device, hpi_handle, row_num, row_data, row_size,error)) {
+			g_warning ("fw meta data write error");
+			return FALSE;
+		}
+
+		/* read flash data */
+		fu_device_set_status (device,FWUPD_STATUS_DEVICE_WRITE);
+
+		g_debug ("Writing Firmware...");
+
+		while (fu_ccgx_cyacd_file_read_row (cyacd_handle, row_buffer,CYACD_ROW_BUFFER_SIZE)) {
+			row_num = *((guint16*)(&row_buffer[0]));
+			row_size = *((guint16*)(&row_buffer[2]));
+			row_data = &row_buffer[4];
+
+			/* write flash data */
+			if (!fu_ccgx_hpi_cmd_write_flash (device, hpi_handle, row_num, row_data, row_size, error)) {
+				g_warning ("fw row data write error at %d row", (gint32)row_num);
+				return FALSE;
+			}
+
+			/* update process */
+			fw_pos = fu_ccgx_cyacd_file_get_pos (cyacd_handle);
+			fu_device_set_progress_full ( device, fw_pos, update_fw_size );
+		}
+
+		/* validate fw */
+		fu_device_set_status(device,FWUPD_STATUS_DEVICE_VERIFY);
+		if (!fu_ccgx_hpi_cmd_validate_fw (device, hpi_handle, update_fw_mode,error)) {
+			g_warning ("fw validate error");
+			return FALSE;
+		}
+
+		fw_pos = fu_ccgx_cyacd_file_get_pos (cyacd_handle);
+		fu_device_set_progress_full (device, fw_pos, update_fw_size);
+
+		break;
+	}
+
+	update_fw_version.val = cyacd_file_info.app_version.val;
+	update_str_version = g_strdup_printf ("%u.%u.%u", 
+						(guint32)update_fw_version.ver.major,
+						(guint32)update_fw_version.ver.minor,
+						(guint32)update_fw_version.ver.build);
+	g_debug ("Update version %s", update_str_version);
+
+	/* update version of device */
+	fu_device_set_version (FU_DEVICE (self), update_str_version);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_dock_bb_write_fw (FuDevice *device,
+			  FuFirmware *firmware,
+			  FwupdInstallFlags flags,
+			  GError **error)
+{
+	FuCcgxDockBb *self = FU_CCGX_DOCK_BB (device);
+	const guint8 *fw_buffer = NULL;
+	gsize fw_size = 0;
+	g_autoptr(GBytes) fw = NULL;
+
+	if (self->dm_device != DM_DEVICE_PD_I2C) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "not supported device type");
+		return FALSE;
+	}
+
+	/* get default Image */
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* get firmware buffer and size */
+	fw_buffer = g_bytes_get_data (fw, &fw_size);
+	if (fw_size <= 0) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "firmware size error");
```

Comment:
You should probably indicate at the end of this the size found wrong (or use g_set_error_literal)

Reply:
OK. I will use g_set_error_literal instead of g_set_error.

Reply:
I fixed it.